File: leaders-nurio-mobile/ios/Routing/OAuthRouteDecisionHandler.swift
import Foundation
@preconcurrency import HotwireNative

@MainActor
final class OAuthRouteDecisionHandler: @preconcurrency RouteDecisionHandler {
    let name = "oauth"

    func matches(location: URL, configuration: Navigator.Configuration) -> Bool {
        SocialAuthRoute.resolve(
            startPath: location.absoluteString,
            baseURL: configuration.startLocation
        ) != nil
    }

    func handle(location: URL, configuration: Navigator.Configuration, navigator: Navigator) -> Router.Decision {
        guard let route = SocialAuthRoute.resolve(
            startPath: location.absoluteString,
            baseURL: configuration.startLocation
        ) else {
            return .cancel
        }

        OAuthSessionCoordinator.shared.presentationAnchorProvider = { [weak navigator] in
            navigator?.activeNavigationController.view.window
        }

        OAuthSessionCoordinator.shared.start(url: route.url) { callbackURL in
            AppRouteCoordinator.shared.handleIncoming(callbackURL)
        }

        return .cancel
    }
}

File: leaders-nurio-mobile/android/app/src/main/java/com/nurio/studyleaders/android/routing/OAuthRouteDecisionHandler.kt
package com.nurio.studyleaders.android.routing

import com.nurio.studyleaders.android.MainActivity
import com.nurio.studyleaders.android.auth.SocialAuthRoute
import dev.hotwire.core.turbo.visit.VisitProposal
import dev.hotwire.navigation.activities.HotwireActivity
import dev.hotwire.navigation.navigator.NavigatorConfiguration
import dev.hotwire.navigation.routing.Router

class OAuthRouteDecisionHandler : Router.RouteDecisionHandler {
    override val name = "oauth-browser-tab"

    override fun matches(
        proposal: VisitProposal,
        configuration: NavigatorConfiguration
    ): Boolean {
        return SocialAuthRoute.resolve(
            proposal.location,
            configuration.startLocation
        ) != null
    }

    override fun handle(
        proposal: VisitProposal,
        configuration: NavigatorConfiguration,
        activity: HotwireActivity
    ): Router.Decision {
        val route = SocialAuthRoute.resolve(
            proposal.location,
            configuration.startLocation
        ) ?: return Router.Decision.CANCEL
        (activity as? MainActivity)?.dispatchSocialAuth(route)

        return Router.Decision.CANCEL
    }
}
